feat: improve text measurements in bound containers

ad1992 opened the pull request:
Adding notes for better clarity

Right now there are inconsistencies since we use a mix of `DOM ` and  `canvas measureText`, this PR targets on removing those inconsistencies and have only one source of truth.

Moving completely to DOM is not the solution -> expensive and will never work on node
Moving to canvas measureText will be faster and also work on node.

## Calculating Width of text element

The support for `canvas.measureText ` is not consistent across browsers try [this](https://128lvh.csb.app/) link in diff browsers 
So therefore I had added an `adjustmentFactor` earlier to accommodate for the inconsistency. I will come back to this one once `height` works well and see if it could be removed

## Calculating height of the text element

Ignore the line height / baseline, that would be covered next.

**Approach #1**

Calculating `height` is possible using canvas metrics [fontBoundingBoxAscent](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/fontBoundingBoxAscent)+[fontBoundingBoxDescent](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/fontBoundingBoxDescent) as it considers font size when calculating the size  and gives same height as DOM element height but it doesn't have support in firefox so need a better heuristic to calculate

The red area is the dimensions by canvas - the height is same as DOM
![image](https://user-images.githubusercontent.com/11256141/216533878-7a9c512d-14c5-46e1-b677-c515d0a92fc1.png)

**Approach #2**
The next thing to try was canvas metrics [actualBoundingBoxAscent](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/actualBoundingBoxAscent) + [actualBoundingBoxDescent](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/actualBoundingBoxDescent) but this doesn't consider font height so doesn't work well with letters which take slightly higher height eg (I)

The red area is the dimensions by canvas
![image](https://user-images.githubusercontent.com/11256141/216534070-56f879bd-4b56-49e3-bd16-b0137237d0de.png)

**Approach #3**
Have a better heuristic for calculating `height` which will  be consistent and work for all browsers.
The DOM element has the property `line-height` which is very imp for spacing between lines. So I am trying to precalulate the line height relative to font size - which we do now as well but with better heuristic -> multiply the font size by 1.2 and when calculating `height` for canvas multiply the recalculated line height by number of lines.
This works well -> will be testing more

![image](https://user-images.githubusercontent.com/11256141/216534930-dde60bee-5641-4aab-9bbb-0c60b6a9bd26.png)

## Baseline
The DOM `baseline` gets removed since now the `render` will be taking care of calculating the positions correctly depending on the canvas `textBaseline`

Since now using canvas textBaseline so calculations depend on the value of `baseline`. 
The `green` line is the x,y coordinate of text element and we need to place it exactly below that line.
I am taking the baseline as `bottom` so the text starts just above the `line` and with the help of line height mentioned above I am pulling it vertically down by `lineHeight` so its exactly below the green line as expected

![image](https://user-images.githubusercontent.com/11256141/216558845-0c13e4a4-be46-4444-8793-65e7309bfef6.png)

<img width="526" alt="Screenshot 2023-02-03 at 2 41 15 PM" src="https://user-images.githubusercontent.com/11256141/216559369-95ea8252-1869-4343-89a5-fdf53337d03a.png">


It can be done with other baseline as well but the calculations will change. 
Removed the vertical offset earlier used in the calculations is not needed anymore. 

## Calculating Width efficiently

### Apporach #1
Coming back to the `adjustment factor` I mentioned earlier. So to remove it we need to make sure the width is consistent across browsers and one way to do that and it works well is by calculating the width of each character, rounding it up and sum it up to get the final width. This works well as you can see the dimensions is same across all browsers and OS [here](https://github.com/excalidraw/excalidraw/pull/6187#issuecomment-1418621324).

So using Math.round has a downside, it doesn't work well when font size increases from small to medium especially since the value of fontSize is increased slightly (16px-20px) and due to rounding off the font size of individual characters remains same eg width of i is ~3.5 when font size is 16px and 4.3 when font size is 20px hence in both case rounding off turns out to be 4px which doesn't go well with a text editor as the width assigned is small then expected.

### Approach #2
Same approach as #1, instead of `Math.round` use `Math.ceil` and this works well for all font sizes, see [here](https://github.com/excalidraw/excalidraw/pull/6187#issuecomment-1418991598).

To test out you can try the [demo](https://uv7pco.csb.app/)

### Apporach #3
Taking each individual character width and using `ceil` could add up to huge discrepancy between canvas and browser measurements eg lets say width of each character in canvas is `1.2px` and we ceil it so its `2px` and we have 10 such characters which means width is `20px` however worst case due to the ceil we have a diff of `0.8*10 = 8px` which could easily fit few chars in text editor and thus increasing discrepancy.  More info in https://github.com/excalidraw/excalidraw/pull/6187#issuecomment-1422351097

Hence to fix this I am taking `ceil` of entire `width` instead so the discrepancy would be way less - though it still will be there due to canvas and browser measurements being different and we will improve this as we see.

And yes with this `adjustment factor` can be removed. To improve perf, we can `cache` the width

TODOS
- [x] Update height when vertical alignment changes to top/bottom -
 fixes https://github.com/excalidraw/excalidraw/issues/6200
- [x] Set the calculated height and stop using `editable.scrollHeight`

## Testing scenarios (copy the diagram in below links (if provided) to preview so test)
This is to make sure nothing existing breaks!

- [x] Basic testing for bound text containers and regular text as well
- [x] Horizontal alignment + edit should work well in all browsers
- [x] Vertical alignment + edit should work well in all browsers

- [x] Adding one letter to https://excalidraw.com/#json=8aAL_CLsjuwOKV4zLrCN0,p5reRLoF7GtLfj3etsol9g should not increase height
- [x] Resize shouldn't increase height https://excalidraw-d9p5r7h9h-excalidraw.vercel.app/#json=hbprXSijRDZ4m3U5BaJ8m,_aQpuJwEwjM_nX9ne19-Sg

- [x] Editor shouldn't autogrow with below steps

open https://excalidraw-hzs43vab7-excalidraw.vercel.app/#json=V8IuuHiLiBN8Tz78LfHFV,O5phO7_M7NSG3kFPpGDZgA

edit

select all

delete

undo

- [x] should not break #5834 
- [x] should not break #6010 
- [x] should not break #5847 
- [x] should not break #5789 
- [x] should not break #6173
- [x] should not break #6137
- [x] should not break #6093
- [x] Collaboration

ad1992 commented:
Coming back to `0.2` adjustment factor which I am currently adding to accommodate for the discrepancies across browsers when using `measureText`, though this seems to work well but here is a better soln.
When calculating `width`, instead of using `measureText` for entire string If I do it by calculating `width` of each char and round it off it works fine and is consistent across the browsers (I am using Mac OS and tested on FF, Chrome and Safari)
I have created a [demo](https://uv7pco.csb.app/) to test out the same with various samples and it works well.

String: `Welcome`

![image](https://user-images.githubusercontent.com/11256141/216907484-cba8688b-2a2a-4906-a19d-cd72bdc7cc4d.png)

@dwelle can you try it out in windows and let me know if the consistency is the same across all browsers?

dwelle commented:
![image](https://user-images.githubusercontent.com/5153846/216933884-ebfe50cd-7b6f-43d8-ab07-557730e52563.png) ![image](https://user-images.githubusercontent.com/5153846/216933898-2763c9bb-cf68-4775-b618-991dd79d873c.png)

ad1992 commented:
> Coming back to `0.2` adjustment factor which I am currently adding to accommodate for the discrepancies across browsers when using `measureText`, though this seems to work well but here is a better soln. When calculating `width`, instead of using `measureText` for entire string If I do it by calculating `width` of each char and round it off it works fine and is consistent across the browsers (I am using Mac OS and tested on FF, Chrome and Safari) I have created a [demo](https://uv7pco.csb.app/) to test out the same with various samples and it works well.
> 
> String: `Welcome`
> 
> ![image](https://user-images.githubusercontent.com/11256141/216907484-cba8688b-2a2a-4906-a19d-cd72bdc7cc4d.png)
> 
> @dwelle can you try it out in windows and let me know if the consistency is the same across all browsers?

Would like to run it in Linux as well, If there is someone who could run in Linux and share the numbers, that would be great!

pavangudiwada commented:
### Details from Linux 

![image](https://user-images.githubusercontent.com/25551553/216941327-df8634fd-a0d6-4a96-8d14-90b13f6d4e6f.png)

ad1992 commented:
> ### Details from Linux
> ![image](https://user-images.githubusercontent.com/25551553/216941327-df8634fd-a0d6-4a96-8d14-90b13f6d4e6f.png)

Thanks for helping in testing out @pavangudiwada, the text "W" is small in "chrome" hence the difference I think, so could you please use the same casing "Welcome" in both browsers and post the screenshot again ?

ad1992 commented:
> Ah, right. Here you go ![image](https://user-images.githubusercontent.com/25551553/216942843-1b1e3c36-d250-407a-b6dd-8a0c620ce2b0.png)

Awesome, so it works the same across all browsers and OS. Thanks once again for helping out @pavangudiwada ❤️

ad1992 commented:
So using `Math.round` has a downside, it doesn't work well when font size increases from small to medium specially since the value of `fontSize` is increased slightly (16px-20px) and due to rounding off the font size of individual characters remains same eg width of `i` is `~3.5` when font size is `16px` and `4.3` when font size is `20px` hence in both case rounding off turns out to be `4px` which doesn't go well with a text editor as the width assigned is small then expected so I am using `Math.ceil` now.

Here's in Mac 👇🏻 

![image](https://user-images.githubusercontent.com/11256141/216968955-c9d60204-59c3-4f95-b299-e4fb09d7552e.png)

So @dwelle @pavangudiwada could you run the [demo](https://uv7pco.csb.app/) once again with text "Excalidraw" and share the screenshots?

pavangudiwada commented:
### Demo from Linux 👇
![image](https://user-images.githubusercontent.com/25551553/216973523-aad0fe97-79d2-4f21-a308-897f4a9f6562.png)

davidluzar commented:
![image](https://user-images.githubusercontent.com/114263536/216974377-f9c9a969-9f8e-4f03-b591-36369594a555.png) ![image](https://user-images.githubusercontent.com/114263536/216974386-b690ce92-1373-47e4-ae11-bf0d41f6bcec.png)

ad1992 commented:
Awesome thanks again for testing @pavangudiwada @davidluzar 💜

DanielJGeiger commented:
@ad1992 I got the following using Firefox and Chromium on Linux:

Firefox:
![Firefox](https://user-images.githubusercontent.com/1852529/217036661-e8379bf8-5bf2-49a8-ab26-70af41f62350.png)

Chromium:
![Chromium](https://user-images.githubusercontent.com/1852529/217036806-99f9cb14-a15b-4da5-b335-b18a0457b3aa.png)

ad1992 commented:
> @ad1992 I got the following using Firefox and Chromium on Linux:
> 
> Firefox: ![Firefox](https://user-images.githubusercontent.com/1852529/217036661-e8379bf8-5bf2-49a8-ab26-70af41f62350.png)
> 
> Chromium: ![Chromium](https://user-images.githubusercontent.com/1852529/217036806-99f9cb14-a15b-4da5-b335-b18a0457b3aa.png)

@DanielJGeiger I am not sure how so much difference is showing up for you and yours is different from @pavangudiwada who is using Linux as well. Which Linux distribution do you have and what versions of ff and chromium?

dwelle commented on src/components/App.tsx:
```diff
@@ -2659,7 +2659,7 @@ class App extends React.Component<AppProps, AppState> {
       // so that the text is centered to cursor position
       if (!parentCenterPosition) {
         mutateElement(element, {
-          y: element.y - element.baseline / 2,
+          y: element.y,
         });
       }
```
this no longer does anything

ad1992 replied:
removed

ad1992 commented:
As per yesterday's discussion with @dwelle - the prev approach works well when it comes to canvas but since I am doing a `Math.ceil` on every character so that difference could eventually add up to a big number leading to increase inconsistency in text editor wrap vs canvas wrap.
Eg lets say width of a char in canvas is 1.2px and browser is 1px and container width is 5px
So for canvas If I have 5 such characters the width will be (2*5 = 10 as I am doing Ceil on every char width) hence it will break to next line as it cross container width (5px)
Where as for browser it will fit in one line as the line width is `(1*5 = 5)`

 So here is a 3rd take to reduce the inconsistency -
 
Instead of `ceil` on every char width, I am just `rounding` off the final width so it's very close to the actual width, and at max, it will have a `diff` of `0.9px` worst case which is not enough to fit any character so early word wrap inconsistency should never happen.

With prev approach (Notice the diff actual metrics vs with ceil)
![image](https://user-images.githubusercontent.com/11256141/216968955-c9d60204-59c3-4f95-b299-e4fb09d7552e.png)

Now
![image](https://user-images.githubusercontent.com/11256141/217497396-3d91c0a6-e809-4f41-8d7d-57d6e071f3d8.png)

@dwelle can you run the [demo](https://uv7pco.csb.app/) in windows  ?

dwelle commented:
![image](https://user-images.githubusercontent.com/5153846/217505129-dddec4d1-7242-44ba-88e4-88c7c48a7a8f.png)

larger string:

![image](https://user-images.githubusercontent.com/5153846/217505109-0ac43658-a656-4975-8bf8-58c95e581b5e.png)

ad1992 commented:
> @ad1992 On further investigation, I doubt the different numbers result from Linux distribution or browser version. The issue appears more along the lines of environment variables or Wayland vs X11. I replicated my previous numbers in the updated sandbox. But then I got the following the other way:
> 
> Firefox: ![image](https://user-images.githubusercontent.com/1852529/217582363-2f0ec8fb-0ad6-4436-a3a2-fda7c9634145.png)
> 
> Chromium: ![image](https://user-images.githubusercontent.com/1852529/217582507-2056acca-121d-4971-91d9-c1b7266f185a.png)

This looks perfect! Though I didn't get what you meant by you were able to replicate the prev numbers in the sandbox ? 
What did you change at your end for the above result ?

ad1992 commented:
Did some more testing to gather more info on discrepancies between text editor and canvas and here is what I found


![image](https://user-images.githubusercontent.com/11256141/218087766-2730fb4c-ba53-4c6d-b7d9-4647b5b0d203.png)

![image](https://user-images.githubusercontent.com/11256141/218087989-c9b0c0e4-ae4b-4f55-9ab0-7a221b7ec26c.png)

long text - 
```
djsd sdasd sdsad asdsdsd sds da dasd sadkljeuqwopocp dasipoqwscascn312903812381290
```
![image](https://user-images.githubusercontent.com/11256141/218088180-a0eb5261-b143-455a-acce-9b4316c16240.png)

As per the testing with a lot of other samples as well, ff seems to take slightly more width than canvas and hence the discrepancy is in ff. All of these cases can be replicated in the preview when you try to fit in one line (irrespective of container width) so I think we should add (~0.05px) when the browser is firefox and see how it goes.

DanielJGeiger commented:
@ad1992 I meant how I had launched the desktop on my computer when I got the original measurements.  I believe that caused the difference.  My original numbers would only indicate issues for some Linux remote desktop scenarios.

ad1992 commented:
> @ad1992 I meant how I had launched the desktop on my computer when I got the original measurements. I believe that caused the difference. My original numbers would only indicate issues for some Linux remote desktop scenarios.

I see ok or it might have been caused by fonts not being loaded as well as that happened to me a couple of times too.

ad1992 commented:
> 

I have added `0.05px` for ff, seems to work well.

ad1992 commented:
So since I am doing `Ceil` its increasing the discrepancies between canvas and text editor in the same browser one such example is 👇🏻 

```
{"type":"excalidraw/clipboard","elements":[{"type":"rectangle","version":506,"versionNonce":489118659,"isDeleted":false,"id":"Zo0jEP8rea20kI0vaFdRJ","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":888.5,"y":372,"strokeColor":"#000000","backgroundColor":"transparent","width":239,"height":298,"seed":969856438,"groupIds":[],"roundness":null,"boundElements":[{"type":"text","id":"moYiAtXSXcYbjxG8B1_2p"}],"updated":1676361267368,"link":null,"locked":false},{"type":"text","version":1362,"versionNonce":811562477,"isDeleted":false,"id":"moYiAtXSXcYbjxG8B1_2p","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":893.5,"y":377,"strokeColor":"#000000","backgroundColor":"transparent","width":229,"height":288,"seed":796592298,"groupIds":[],"roundness":null,"boundElements":[],"updated":1676361267368,"link":null,"locked":false,"fontSize":20,"fontFamily":1,"text":"asd asd asdasd; \nalksd; la;dl \na;sdlsdla;sd \najfkjawrfioajwfjiawmcuo\nm2j0cj208 j80cj \nw08jc8024j cj2804j \n8024jc082j480c \nj2480jc \n8024j824j999 k k k k\nk k kk a ll la lal l \nallalalall al lla ll \nalsllll","textAlign":"center","verticalAlign":"middle","containerId":"Zo0jEP8rea20kI0vaFdRJ","originalText":"asd asd asdasd; alksd; la;dl a;sdlsdla;sd ajfkjawrfioajwfjiawmcuo m2j0cj208 j80cj w08jc8024j cj2804j 8024jc082j480c j2480jc 8024j824j999 k k k k k k kk a ll la lal l allalalall al lla ll alsllll"}],"files":{}}
```
This gets fixed If I stop Ceil or take up to a precision

So we have 3 options 

| Approach |. Discrepancies between text editor and canvas in same browser | Discrepancies cross browser |
| ---- | ------- | ---- |
| No Ceil | Least | Highest |
| Ceil with Precision | Medium | Medium |
| Ceil without precision | Highest | Least |

As mentioned earlier I was trying to solve cross browser due to which I was taking `ceil` - reason being so that behaviour is persistent when copy pasted cross browser and collaboration.
So keeping that aside If we look at same browser behaviour than `Taking No Ceil` seems to be the best option and that's what I have pushed in my last commit so that in same browser the behaviour is consistent.

Note that text editor and browser discrepancies still might be there due to the API's being different and for that we need to adjust the dimensions of text editor (eg firefox so now I am adding 0.5px as I am not using ceil)

Coming back to cross browser since that seems to be highest in this approach ->
When copy pasting cross browser the dimensions might change (When you try to edit/refresh) but post that it would get adjusted to the canvas dimensions.
eg try pasting this in chrome and notice the change after refresh, however in firefox it will be same after refresh as its pasted from firefox.

```
{"type":"excalidraw/clipboard","elements":[{"type":"rectangle","version":606,"versionNonce":1619729716,"isDeleted":false,"id":"21cjg9ReflUuUDdcHxBgI","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":1106.4285150216242,"y":1946.8100108037324,"strokeColor":"#000000","backgroundColor":"transparent","width":46,"height":82,"seed":937152459,"groupIds":[],"roundness":{"type":3},"boundElements":[{"type":"text","id":"uDZaOYMAlU7-rMjzu2umH"}],"updated":1676353445992,"link":null,"locked":false},{"type":"text","version":404,"versionNonce":388464652,"isDeleted":false,"id":"uDZaOYMAlU7-rMjzu2umH","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":1111.4285150216242,"y":1978.2100108037325,"strokeColor":"#000000","backgroundColor":"transparent","width":36,"height":19.2,"seed":543303269,"groupIds":[],"roundness":null,"boundElements":[],"updated":1676353445992,"link":null,"locked":false,"fontSize":16,"fontFamily":1,"text":"shirt","textAlign":"center","verticalAlign":"middle","containerId":"21cjg9ReflUuUDdcHxBgI","originalText":"shirt"}],"files":{}}
```
Lets come to collaboration for the same scenario - So if it happens that dimensions are different (as in above case) then if you try to edit the above in chrome -> The width changes and so does the text (text will become "shir\nt") and same is affected in all. browsers so ff in spite of it can fit in one line will start showing in two lines and hence it will be same across all browsers. But yes frequent edits to this kind of test cases could lead to more update to dimensions (chrome edits -> ff edits -> chrome edits -> and so on for same element).
And since in prod we are using DOM measurements in text editor in measureText in wrap so these kind of cases would be there in prod as well even now.
As we observe more we can improve this cross-browser. 
cc / @dwelle 

@DanielJGeiger would be great if you could test it out as well and let me know if you come across anything which works in prod but breaks in my PR 💜

ad1992 commented:
> So since I am doing `Ceil` its increasing the discrepancies between canvas and text editor in the same browser one such example is 👇🏻
> 
> ```
> {"type":"excalidraw/clipboard","elements":[{"type":"rectangle","version":506,"versionNonce":489118659,"isDeleted":false,"id":"Zo0jEP8rea20kI0vaFdRJ","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":888.5,"y":372,"strokeColor":"#000000","backgroundColor":"transparent","width":239,"height":298,"seed":969856438,"groupIds":[],"roundness":null,"boundElements":[{"type":"text","id":"moYiAtXSXcYbjxG8B1_2p"}],"updated":1676361267368,"link":null,"locked":false},{"type":"text","version":1362,"versionNonce":811562477,"isDeleted":false,"id":"moYiAtXSXcYbjxG8B1_2p","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":893.5,"y":377,"strokeColor":"#000000","backgroundColor":"transparent","width":229,"height":288,"seed":796592298,"groupIds":[],"roundness":null,"boundElements":[],"updated":1676361267368,"link":null,"locked":false,"fontSize":20,"fontFamily":1,"text":"asd asd asdasd; \nalksd; la;dl \na;sdlsdla;sd \najfkjawrfioajwfjiawmcuo\nm2j0cj208 j80cj \nw08jc8024j cj2804j \n8024jc082j480c \nj2480jc \n8024j824j999 k k k k\nk k kk a ll la lal l \nallalalall al lla ll \nalsllll","textAlign":"center","verticalAlign":"middle","containerId":"Zo0jEP8rea20kI0vaFdRJ","originalText":"asd asd asdasd; alksd; la;dl a;sdlsdla;sd ajfkjawrfioajwfjiawmcuo m2j0cj208 j80cj w08jc8024j cj2804j 8024jc082j480c j2480jc 8024j824j999 k k k k k k kk a ll la lal l allalalall al lla ll alsllll"}],"files":{}}
> ```
> 
> This gets fixed If I stop Ceil or take up to a precision
> 
> So we have 3 options
> 
> Approach	. Discrepancies between text editor and canvas in same browser	Discrepancies cross browser
> No Ceil	Least	Highest
> Ceil with Precision	Medium	Medium
> Ceil without precision	Highest	Least
> As mentioned earlier I was trying to solve cross browser due to which I was taking `ceil` - reason being so that behaviour is persistent when copy pasted cross browser and collaboration. So keeping that aside If we look at same browser behaviour than `Taking No Ceil` seems to be the best option and that's what I have pushed in my last commit so that in same browser the behaviour is consistent.
> 
> Note that text editor and browser discrepancies still might be there due to the API's being different and for that we need to adjust the dimensions of text editor (eg firefox so now I am adding 0.5px as I am not using ceil)
> 
> Coming back to cross browser since that seems to be highest in this approach -> When copy pasting cross browser the dimensions might change (When you try to edit/refresh) but post that it would get adjusted to the canvas dimensions. eg try pasting this in chrome and notice the change after refresh, however in firefox it will be same after refresh as its pasted from firefox.
> 
> ```
> {"type":"excalidraw/clipboard","elements":[{"type":"rectangle","version":606,"versionNonce":1619729716,"isDeleted":false,"id":"21cjg9ReflUuUDdcHxBgI","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":1106.4285150216242,"y":1946.8100108037324,"strokeColor":"#000000","backgroundColor":"transparent","width":46,"height":82,"seed":937152459,"groupIds":[],"roundness":{"type":3},"boundElements":[{"type":"text","id":"uDZaOYMAlU7-rMjzu2umH"}],"updated":1676353445992,"link":null,"locked":false},{"type":"text","version":404,"versionNonce":388464652,"isDeleted":false,"id":"uDZaOYMAlU7-rMjzu2umH","fillStyle":"hachure","strokeWidth":1,"strokeStyle":"solid","roughness":1,"opacity":100,"angle":0,"x":1111.4285150216242,"y":1978.2100108037325,"strokeColor":"#000000","backgroundColor":"transparent","width":36,"height":19.2,"seed":543303269,"groupIds":[],"roundness":null,"boundElements":[],"updated":1676353445992,"link":null,"locked":false,"fontSize":16,"fontFamily":1,"text":"shirt","textAlign":"center","verticalAlign":"middle","containerId":"21cjg9ReflUuUDdcHxBgI","originalText":"shirt"}],"files":{}}
> ```
> 
> Lets come to collaboration for the same scenario - So if it happens that dimensions are different (as in above case) then if you try to edit the above in chrome -> The width changes and so does the text (text will become "shir\nt") and same is affected in all. browsers so ff in spite of it can fit in one line will start showing in two lines and hence it will be same across all browsers. But yes frequent edits to this kind of test cases could lead to more update to dimensions (chrome edits -> ff edits -> chrome edits -> and so on for same element). And since in prod we are using DOM measurements in text editor in measureText in wrap so these kind of cases would be there in prod as well even now. As we observe more we can improve this cross-browser. cc / @dwelle
> 
> @DanielJGeiger would be great if you could test it out as well and let me know if you come across anything which works in prod but breaks in my PR 💜

added `0.05px` in Safari too for text editor as now using `ceil` its needed in safari as well.

DanielJGeiger commented:
@ad1992 I did not find this PR breaking anything from production.

ad1992 commented:
> @ad1992 I did not find this PR breaking anything from production. Several fixes verified as well.

Awesome thanks for testing it out @DanielJGeiger 💜

ad1992 commented:
Alright merged 🚀
Thanks everyone for helping out in testing 😀

dobegor commented:
Hey @ad1992!
I recently found this when I stumbled upon an unfortunate regression for some of my existing diagrams in Excalidraw.
Here's the "before and after".

Before (this is the screenshot I took with an older version of Excalidraw):
![CleanShot 2023-04-26 at 13 39 55](https://user-images.githubusercontent.com/14029650/234563956-2212a557-b7cf-40da-82f8-addd1a6d8467.png)

After (this is how it looks now with the same diagram):
![CleanShot 2023-04-26 at 13 40 55](https://user-images.githubusercontent.com/14029650/234564300-d236a7ee-b4a5-4d6d-97a7-489d30692298.png)

For some reason, the text got wrapped in some blocks. Some of them have enough place to display text without wrapping.
If I start editing the node's text, the following happens:

Before clicking:
![CleanShot 2023-04-26 at 13 42 48](https://user-images.githubusercontent.com/14029650/234564623-3fb2efa6-064a-400b-9c6a-3fb2a4548b30.png)
After clicking:
![CleanShot 2023-04-26 at 13 42 54](https://user-images.githubusercontent.com/14029650/234564651-35fe1763-ae06-497f-b3f8-59a6342d44d9.png)

Another example:
Before I click:
![CleanShot 2023-04-26 at 13 43 28](https://user-images.githubusercontent.com/14029650/234564773-df26faca-2611-4fe2-a95c-7e9679e4dcbe.png)
After:
![CleanShot 2023-04-26 at 13 43 28](https://user-images.githubusercontent.com/14029650/234564807-5ff9da4e-e61b-4bd3-a4fd-4c6ded03fd3f.png)

It looks like it could have broken a lot of existing diagrams.
But also, I can no longer make diamond-shaped nodes with text within a reasonable size.
For example, this fitted perfectly fine before:
![CleanShot 2023-04-26 at 13 45 09](https://user-images.githubusercontent.com/14029650/234565121-b0f07e70-ae38-4262-9913-0dcb078d2be3.png)

But the smallest I can get it to looks like this now:
![CleanShot 2023-04-26 at 13 45 37](https://user-images.githubusercontent.com/14029650/234565325-5ba8fb44-78ac-4882-bb76-80585c267fdf.png)

Is there any way to address this?
Thank you!

ad1992 commented:
Hey @dobegor so sorry about the regression in the diagrams, I think the regression was introduced in https://github.com/excalidraw/excalidraw/pull/6172 (where we fixed the ellipse bounding box) and https://github.com/excalidraw/excalidraw/pull/6265 (where we fixed the diamond bounding box), can you share a sample diagram in older version which I can use for testing and I will surely try to get it to the state it was before.

dobegor commented:
@ad1992 sure, let's continue the conversation in #6523
